storage/parquet_shard.go

```diff
@@ -26,77 +26,88 @@ import (
 	"github.com/prometheus-community/parquet-common/schema"
 )
 
-var DefaultShardOptions = shardOptions{
-	optimisticReader: true,
+var DefaultShardOptions = ShardOptions{
+	OptimisticReader:           true,
+	PagePartitioningMaxGapSize: 10 * 1024,
 }
 
-type shardOptions struct {
-	fileOptions      []parquet.FileOption
-	optimisticReader bool
+type ShardOptions struct {
+	FileOptions                []parquet.FileOption
+	OptimisticReader           bool
+	PagePartitioningMaxGapSize int
 }
 
 type ParquetFile struct {
 	*parquet.File
 	ReadAtWithContextCloser
 	BloomFiltersLoaded bool
 
-	optimisticReader bool
+	optimisticReader           bool
+	pagePartitioningMaxGapSize int
 }
 
-type ShardOption func(*shardOptions)
+type ShardOption func(*ShardOptions)
 
 func WithFileOptions(fileOptions ...parquet.FileOption) ShardOption {
-	return func(opts *shardOptions) {
-		opts.fileOptions = append(opts.fileOptions, fileOptions...)
+	return func(opts *ShardOptions) {
+		opts.FileOptions = append(opts.FileOptions, fileOptions...)
 	}
 }
 
 func WithOptimisticReader(optimisticReader bool) ShardOption {
-	return func(opts *shardOptions) {
-		opts.optimisticReader = optimisticReader
+	return func(opts *ShardOptions) {
+		opts.OptimisticReader = optimisticReader
 	}
 }
 
-func (f *ParquetFile) GetPages(ctx context.Context, cc parquet.ColumnChunk, pagesToRead ...int) (*parquet.FilePages, error) {
+// WithPageMaxGapSize set the max gap size between pages that should be downloaded together in a single read call
+func WithPageMaxGapSize(pagePartitioningMaxGapSize int) ShardOption {
+	return func(opts *ShardOptions) {
+		opts.PagePartitioningMaxGapSize = pagePartitioningMaxGapSize
+	}
+}
+
+func (f *ParquetFile) GetPages(ctx context.Context, cc parquet.ColumnChunk, minOffset, maxOffset int64) (*parquet.FilePages, error) {
 	colChunk := cc.(*parquet.FileColumnChunk)
 	reader := f.WithContext(ctx)
 
-	if len(pagesToRead) > 0 && f.optimisticReader {
-		offset, err := cc.OffsetIndex()
-		if err != nil {
-			return nil, err
-		}
-		minOffset := offset.Offset(pagesToRead[0])
-		maxOffset := offset.Offset(pagesToRead[len(pagesToRead)-1]) + offset.CompressedPageSize(pagesToRead[len(pagesToRead)-1])
-		reader = newOptimisticReaderAt(reader, minOffset, maxOffset)
+	if f.optimisticReader {
```

Comment:
Could we drop a docstring somewhere just generally describing what the behavior of the vanilla vs. optimistic readers are here, or why we might configure one or the other?

Reply:
Merged then in a single configuration.